File: packages/guards/src/profanity-hi.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ProfanityHiOptions {
  action: 'block' | 'warn';
}

const WORDS: string[] = [
  'बकवास', 'भड़वा', 'हरामी', 'कुत्ता', 'गधा',
  'चूतिया', 'मादरचोद', 'बहनचोद', 'लौड़ा', 'गांड',
];

export function profanityHi(options: ProfanityHiOptions): Guard {
  return {
    name: 'profanity-hi',
    version: '0.1.0',
    description: 'Hindi profanity detection',
    category: 'locale',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const word of WORDS) {
        if (text.includes(word)) matched.push(word);
      }

      const triggered = matched.length > 0;

      return {
        guardName: 'profanity-hi',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: triggered ? Math.min(matched.length / 3, 1.0) : 0,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matched } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/profanity_hi.py
"""Detect Hindi profanity."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_WORDS = [
    "बकवास", "भड़वा", "हरामी", "कुत्ता",
    "गधा", "चूतिया", "मादरचोद", "बहनचोद",
    "लौड़ा", "गांड",
]

_PATTERNS = [
    re.compile(rf"(?:^|\s){re.escape(w)}(?:\s|$|[।,!?.])", re.IGNORECASE)
    for w in _WORDS
]


class _ProfanityHi:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "profanity-hi"
        self.action = action

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for pat in _PATTERNS:
            if pat.search(text):
                matched.append(pat.pattern)

        triggered = len(matched) > 0
        score = min(len(matched) / 3, 1.0) if triggered else 0.0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="profanity-hi",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Hindi profanity detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "matched_count": len(matched),
                    "reason": (
                        "Text contains Hindi profane"
                        " or offensive words"
                    ),
                }
                if triggered
                else None
            ),
        )


def profanity_hi(
    *, action: str = "block"
) -> _ProfanityHi:
    return _ProfanityHi(action=action)
